### PR TITLE
Ast collector fixes

### LINKED ISSUE #7: Enhance AST dump
I'm working on an experimental branch to enhance `-frust-dump-parse` which is a little inspired by gccgo's AST dump.
For example:
```rust
 fn abc(x:u32, y:u32) -> u32 {
    return x+y;
 }
 
 fn main() {
     println!("Hello World!");
 }
```
The dumped result looks like this:
```
Crate: 
 inner attributes: none
 items: 
  
u32 abc(x : u32, y : u32)
 {
 BlockExpr: 
 outer attributes: none
 inner attributes: none
 statements: 
  ExprStmtWithoutBlock: return ArithmeticOrLogicalExpr: x + y
 final expression: none
 }

  
void main()
 {
 BlockExpr: 
 outer attributes: none
 inner attributes: none
 statements: 
  ExprStmtWithoutBlock: println!("Hello World!")
 final expression: none
 }
```

The branch is [nala/enhance/ast-dump](https://github.com/philberty/gccrs/tree/nala/enhance/ast-dump)

Comments are welcome.

### PR DESCRIPTION
**Pull request by P-E-P**
Fixes several issues with current ast collector.

Requires #2313 

Fixes #7

### CODE REVIEW

**Comment by P-E-P** [comment]
I believe the last remaining issues to compile properly exists because of #2126

File: gcc/rust/ast/rust-ast-collector.cc [hunk 1]
```diff
@@ -2200,22 +2194,33 @@ TokenCollector::visit (ExternalStaticItem &item)
...
+  if (function.has_generics ())
+    visit (function.get_generic_params ());
   push (Rust::Token::make (LEFT_PAREN, Location ()));
 
   visit_items_joined_by_separator (function.get_function_params ());
+  if (function.is_variadic ())
+    {
+      push (Rust::Token::make (COMMA, Location ()));
+      // TODO: Add variadic outer attributes?
+      // TODO: Add variadic name once implemented.
```
Comment (CohenArthur) [inline gcc/rust/ast/rust-ast-collector.cc:2213]: what's a variadic name? `...args`?

**Review by CohenArthur — APPROVED** [review]
Thank you! Looks good to me :)

File: gcc/rust/ast/rust-ast-collector.cc (hunk 1 above)
Comment (CohenArthur) [reply gcc/rust/ast/rust-ast-collector.cc:2213]: thanks!

**Comment by P-E-P** [comment]
@philberty Should this issue close #7 ?

**Comment by CohenArthur** [comment]
@P-E-P yes, I think this should close #7. The AST dump will now be in a very good state compared to before. Edit your original message and I'll merge this PR

**Comment by P-E-P** [comment]
> @P-E-P yes, I think this should close #7. The AST dump will now be in a very good state compared to before. Edit your original message and I'll merge this PR

Done!

**Pull request merged**